OpenSim/Common/Delay.h

```diff
@@ -0,0 +1,179 @@
+ *
+ * For times prior to the start of a simulation this Component behaves as
+ * though the input value had been constant at its initial value.
+ *
+ * @tparam T The type of the quantity to be delayed. Common choices are a
+ * SimTK::Real (see ScalarDelay) or a SimTK::Vector (see VectorDelay).
+ *
+ * @see ScalarDelay, VectorDelay
+ *
+ *
+ * This Component has a single input named "input" and a single output named
+ * "output". The typical way to use this Component is as a subcomponent in
+ * another component. Here is a basic example in which a controller specifies
+ * the control signal for a single actuator using the delayed value of a
+ * coordinate:
+ * @code
+ * using namespace OpenSim;
+ * using namespace SimTK;
+ * class MyController : Controller {
+ * OpenSim_DECLARE_CONCRETE_OBJECT(MyController, Controller);
+ * public:
+ *     OpenSim_DECLARE_PROPERTY(delay, double, "Duration of delay (seconds).");
+ *     MyController() { constructProperties(); }
+ *     void computeControls(const State& s, Vector &controls) const override {
+ *         double q = _coordDelay.getValue(s);
+ *         // Delayed negative feedback control law.
+ *         double u = -10 * q;
+ *         getModel().getActuatorSet()[0].addInControls(Vector(1, u), controls);
+ *     }
+ * private:
+ *     void constructProperties() {
+ *          constructProperty_delay(0.01); // 10 milliseconds
+ *     }
+ *     void extendFinalizeFromProperties() override {
+ *         _coordDelay.set_delay(get_delay());
+ *         addComponent(&_coordDelay);
+ *     }
+ *     void extendConnectToModel(Model& model) override {
+ *         Super::extendConnectToModel(model);
+ *         const auto& coord = model.getCoordinateSet()[0];
+ *         // Set the input for the Delay component.
+ *         _coordDelay.getInput("input").connect(coord.getOutput("value"));
+ *     }
+ *
+ *     ScalarDelay _coordDelay;
+ * };
+ * @endcode
+ *
+ * This class is implemented via a SimTK::Measure_<T>::Delay.
+ *
+ * @ingroup operators
+ *
+ */
+template<typename T>
+class Delay : public Component {
+OpenSim_DECLARE_CONCRETE_OBJECT_T(Delay, T, Component);
+public:
+
+    /** @name Property declarations
+    These are the serializable properties associated with a Delay. **/
+    /**@{**/
+    OpenSim_DECLARE_PROPERTY(delay, double,
+        "The duration (nonnegative, seconds) by which the output is delayed.");
+    /**@}**/
+
+    /// Default constructor.
+    Delay();
+
+    /// Convenience constructor that sets the delay property.
+    explicit Delay(double delay);
+
+    /// Get the delayed value (the input's value at time t-delay).
+    T getValue(const SimTK::State& s) const;
+
+private:
+
+    void constructProperties() override;
+    void constructInputs() override;
+    void constructOutputs() override;
+
+    void extendFinalizeFromProperties() override;
+    void extendAddToSystem(SimTK::MultibodySystem& system) const override;
+
+    SimTK::MeasureIndex _delayMeasureIndex;
```

Comment:
> For related discussion, see #343

That linked to an OpenSim issue; I think you meant simbody/simbody#343.

> Would ClearOnCopy be a smart pointer? In Simbody?

I was just thinking out loud so this might not make sense. But I wasn't thinking of a pointer at all but more like a "mix in" template class that would replace a class's copy constructor with its default constructor and its copy assignment with a call to its clear() method. The data member could then be:
```c++
ClearOnCopy<SimTK::Measure> m_delayMeasure;
```
The implementation could be something like
```c++
template <class T> ClearOnCopy : public T {
    ClearOnCopy(const ClearOnCopy&) : T() {} // copy constructor does default construct
    ClearOnCopy& operator=(const ClearOnCopy&) {clear();} // copy assignment
    // or maybe:
    ClearOnCopy& operator=(const ClearOnCopy&) {
        static_cast<T>(*this) = T();
    }
};
```

I'm not sure that would be enough to make it useful though. Anything that would achieve the goal of not requiring a copy constructor to be written would be good.